Client/mods/deathmatch/logic/CClientIFP.cpp

```diff
@@ -0,0 +1,1079 @@
+#include <StdInc.h>
+#include "game/CAnimBlendSequence.h"
+#include "game/CAnimBlendHierarchy.h"
+
+CClientIFP::CClientIFP(class CClientManager* pManager, ElementID ID) : CClientEntity(ID)
+{
+    // Init
+    m_pManager = pManager;
+    SetTypeName("IFP");
+    m_pIFPAnimations = std::make_shared<CIFPAnimations>();
+    m_pAnimManager = g_pGame->GetAnimManager();
+    m_bVersion1 = false;
+    m_bUnloading = false;
+    m_u32Hashkey = 0;
+}
+
+bool CClientIFP::LoadIFP(const SString& strFilePath, const SString& strBlockName)
+{
+    m_strBlockName = strBlockName;
+    m_pVecAnimations = &m_pIFPAnimations->vecAnimations;
+
+    if (LoadIFPFile(strFilePath))
+    {
+        m_u32Hashkey = HashString(strBlockName.ToLower());
+        return true;
+    }
+    return false;
+}
+
+bool CClientIFP::LoadIFPFile(const SString& strFilePath)
+{
+    if (LoadFileToMemory(strFilePath))
+    {
+        char Version[4];
+        ReadBytes(Version, sizeof(Version));
+
+        bool bAnp3 = strncmp(Version, "ANP3", sizeof(Version)) == 0;
+        bool bAnp2 = strncmp(Version, "ANP2", sizeof(Version)) == 0;
+        if (bAnp2 || bAnp3)
+        {
+            m_bVersion1 = false;
+            ReadIFPVersion2(bAnp3);
```

Comment:
What if the IFP file is corrupt? Shouldn't this function call return a boolean to indicate failure? See also ReadIFPVersion1

Reply:
This is how GTA:SA was reading IFP files. It's easy to check for corruption in case of DFF, it has multiple headers throughout the file, but IFP has only 1 header which is at the top, we can use the first 8 bytes to check:
```struct SIFPHeader 
{
    char         Version [ 4 ]; 
    std::int32_t EndOfFileOffset; 
};
``` 
We can add two checks:

**Version**
It can have three possible strings, "ANP1," "ANP2," and "ANP3," if we don't find these strings, we can assume that it's corrupt, and return false. 

**EndOfFileOffset**
We can do ```LengthOfFile=== EndOfFileOffset-8bytes``` If that's not the case, we can assume it's corrupt. 

The client can still crash, if the header is fine, but the data inside the IFP file is corrupt. Nonetheless, Adding these two checks will certainly help.

Reply:
Looks like IFP files modified with KAM's script have incorrect EndOfFileOffset, so we cannot rely on that, I just confirmed it. We can only check the first four bytes.